[87] Copy `NodeStateAdaptor` from the wallet to the node

Anviking opened the pull request:
## Description
We don't want to use NodeStateAdaptor from the wallet. We'd like to decouple the wallet from the node, and talk over an API (the Node API). To aid the transition, I have abruptly copied the wallet `NodeStateAdaptor` to the Node, so that it can serve the following additional protocol parameters in the node-settings endpoint:
- slotId
- slotCount
- maxTxSize
- feePolicy
- securityParameter
## Linked issue

input-output-hk/cardano-wallet#87

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [x] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub. Nope

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [ ] I have added tests to cover my changes. No
- [x] All new and existing tests passed. Sure

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->

```fish
cd cardano-sl

stack build cardano-sl-node:cardano-node-simple --ghc-options=-optl-Wl,-dead_strip_dylibs

stack exec cardano-node-simple -- --topology=wallet/topology-examples/testnet.yaml \
                             --configuration-key=mainnet_staging_short_epoch_full \
                             --log-config=log-configs/cluster.yaml

curl https://localhost:8080/api/v1/node-settings \
   --cacert scripts/tls-files/ca.crt \
   --cert scripts/tls-files/client.pem | jq .
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--100   464    0   464    0     0   7708      0 --:--:-- --:--:-- --:--:--  7733
{
  "data": {
    "slotId": {
      "siEpoch": 0,
      "siSlot": {
        "getSlotIndex": 0
      }
    },
    "slotDuration": {
      "quantity": 20000,
      "unit": "milliseconds"
    },
    "slotCount": 120,
    "softwareInfo": {
      "version": 0,
      "applicationName": "cardano-sl"
    },
    "projectVersion": "2.0.0",
    "gitRevision": "4fe6e653ac6b4b70ae629133a8eef2e364b3a4fd",
    "maxTxSize": 4096,
    "feePolicy": {
      "txSizeLinear": {
        "a": 155381,
        "b": 43.946
      }
    },
    "securityParameter": 12
  },
  "status": "success",
  "meta": {
    "pagination": {
      "totalPages": 1,
      "page": 1,
      "perPage": 1,
      "totalEntries": 1
    }
  }
}
```

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

## How to merge

Send the message `bors r+` to merge this PR. For more information, see
[`docs/how-to/bors.md`](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/how-to/bors.md).

Anviking commented on node/src/Cardano/Node/API.hs:
```diff
@@ -219,6 +232,27 @@ instance Core.HasSlottingVar SettingsCtx where
+                <*> (getMaxTxSize ns)
+                <*> (getFeePolicy ns)
+                <*> (getSecurityParameter ns)
+                <*> (getSlotCount ns)
```
This is the essential part

Anviking commented on lib/src/Pos/Node/API.hs:
```diff
@@ -452,6 +453,10 @@ newtype V1 a = V1 a deriving (Eq, Ord)
 unV1 (V1 a) = a
 
+
+instance Arbitrary SlotDuration where
```
Had to move this instance a bit to avoid TH complaining

Anviking commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +551,130 @@ instance ToSchema (V1 Version) where
+deriving instance FromJSON (V1 Core.SlotId)
+
+instance Arbitrary (V1 Core.SlotId) where
+    arbitrary = fmap V1 arbitrary
```
> Do you actually need those instances? If it's about the Example type-class (which default to using an Arbitrary instance if any), you better go defining an example by hand.

— @KtorZ (https://github.com/input-output-hk/cardano-sl/pull/3971/files#r242542685)

Anviking replied:
Think they're only used for the `Example` for `NodeSettings`. If so: are you sure?

KtorZ replied:
Try to remove them, look who's complaining :upside_down_face:

Anviking replied:
Not possible. `Arbitrary` is a superclass of `Example`.
https://github.com/input-output-hk/cardano-sl/blob/5ecc6f58e6b4e0a92bbca696824f027387db6d86/lib/src/Pos/Util/Example.hs#L18-L20

We use `Example` here: (Which I guess could be changed, but that feel out of scope of this PR) https://github.com/input-output-hk/cardano-sl/blob/5ecc6f58e6b4e0a92bbca696824f027387db6d86/lib/src/Pos/Node/API.hs#L65

KtorZ replied:
Ah! True we did constrain every `Example` with `Arbitrary`. Sad. Let's not bother with that in this ticket then.

KtorZ commented:
Gotta get rid of the changes made to `/wallet` as well, these should only happen in `cardano-wallet` :+1:

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +551,130 @@ instance ToSchema (V1 Version) where
+    , setSoftwareInfo      :: !(V1 Core.SoftwareVersion)
+    , setProjectVersion    :: !(V1 Version)
+    , setGitRevision       :: !Text
+    , setMaxTxSize         :: !Byte
```
A bit rough at the API-level. A value with units would be a bit more frirendly (cf `SlotDuration` for instance)

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +551,130 @@ instance ToSchema (V1 Version) where
+instance ToSchema (V1 Core.TxFeePolicy) where
+    declareNamedSchema _ =
+        pure $ NamedSchema (Just "Core.TxFeePolicy") $ mempty
+            & type_ .~ SwaggerString
```
I see this is represented as:

```json
    "feePolicy": {
      "txSizeLinear": {
        "a": 155381,
        "b": 43.946
      }
    }
```

So, not a string, but rather an object with a few properties ;)

Anviking commented:
@KtorZ These should be required changes for compiling.

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +551,130 @@ instance ToSchema (V1 Version) where
+            & type_ .~ SwaggerString
+
+deriving instance ToJSON (V1 Core.TxFeePolicy)
+deriving instance FromJSON (V1 Core.TxFeePolicy)
```
The derived instances are quite unsatisfactory here :/ ... we end up with a `txSizeLinear` whereas `linear` or `sizeLinear` would be nicer.

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +551,130 @@ instance ToSchema (V1 Version) where
+    , setGitRevision       :: !Text
+    , setMaxTxSize         :: !Byte
+    , setFeePolicy         :: !(V1 Core.TxFeePolicy)
+    , setSecurityParameter :: !SecurityParameter
```
Same remark here, what's the unit of that thing :thinking: ?

Anviking replied:
If anything: slots. But technically, no unit, I believe 🤔 

"k slots"
"k \in \mathbb{Z}"

https://cardanodocs.com/technical/
https://eprint.iacr.org/2016/889.pdf

KtorZ replied:
Well, `slots` sounds like a very valid unit to me then ^^"

Anviking replied:
> Well, slots sounds like a very valid unit to me then ^^"

But… If `k = 2 slots` then `k slots = 2 slots^2` 😨

Are we looking past this and redefining (in a minor way) `k` for the sake of clarity in the API?

Anviking replied:
I think this is the simplest form of my reasoning:

In the docs/paper we can find the expressions like `k: 60` (as in `k = 60`), and `k slots`.

Assume our API instead returns `k = 2 slots`, then:
```haskell
(let k = 2 slots  -- From our API
in k slots       -- we see this expression in the paper)

== 2 slots^2     -- this is absurd unless slots^2 == slots == 1, i.e has no dimension
```

Without a unit in our API we get this:
```haskell
(let k = 2       -- From our API
in k slots       -- we see this expression in the paper)

== 2 slots -- Very sane
```

**I'd be very happy to be wrong or to ignore this as this is fairly non-important in comparison to other things.**

Reasons I see I could be wrong:
- `k slots` in the text is meant more as "a distance d" rather than "d * meters"
- `slots = 1`
- We don't care about dimensional analysis here / it's unsuitable here ?
- ?

But maybe @paweljakubas could weigh in, who should know a lot more than me? Should `setSecurityParameter`, *k*, be described like `{"unit": "slots", quantity: 2}` or `2`  in the API?

Anviking replied:
Actually — the simplest version of the argument is more trivial 😅: 

If the docs say `k = 60`, *k* doesn't have a dimension (and we can't say `k = 60 slots`).

Anviking replied:
> we refer to k as let k = 2 in k slots, thus, the unit for securityParameter is slot. k refer to a number of slots... ?

@KtorZ I think you are saying that `securityParameter = k slots`. However, in the paper we can read: "some security parameter k", meaning `securityParameter = k`.

Anviking commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +553,209 @@ instance ToSchema (V1 Version) where
+        case tag of
+            "linear" ->
+                Core.TxFeePolicyTxSizeLinear <$> parseJSON j
+            "unknown" -> do
```
*If* possible there would be an opportunity here, I think, as we're already defining our own API specific types (currently `V1 Core.TxFeePolicy`), different from both `Core.TxFeePolicy` and the JSON representation.

KtorZ replied:
I think that makes total sense. Having an API representation that we map from and to `TxFeePolicy` :+1:

Anviking replied:
So generics + specific types? I should have asked about this, sorry — thought we would have needed something much more aircastel-y, like DerivingVia, or something. I decided to stick with the handwritten instances to avoid unpleasant surprises for the time being.

I'm still not fully convinced there aren't any unpleasant surprises if we were to use generics instead. Are there?

My quick take:
- We need mapping both ways / isomorphism between `Core.SomeType` and `Node.API.SomeType`
- We would need to duplicate hierarchies of types [Core.TxFeePolicy](https://github.com/input-output-hk/cardano-sl/blob/develop/core/src/Pos/Core/Common/TxFeePolicy.hs#L44) so that the two `Coeff` can be replaced with `MeasuredIn 'AdaPerByte Coeff` and `MeasuredIn 'Ada Coeff`.
- We need Aeson + ToSchema instances for `MeasuredIn` (that's basically possible I believe?) Though might lose some flexibility (we can't provide the name in `Aeson.withObject "name"`)

Anviking replied:
Might explore
```haskell
instance (ToJSON q, UnitToText u) => ToJSON (MeasuredIn (u :: UnitOfMeasure) q)`
```
instead of current helper-functions, in a separate PR.

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +553,209 @@ instance ToSchema (V1 Version) where
+        pure $ NamedSchema (Just "SecurityParameter") $ mempty
+            & type_ .~ SwaggerNumber
+            & maximum_ .~ Just (fromIntegral (maxBound :: Int))
+            & minimum_ .~ Just (fromIntegral (minBound :: Int))
```
If the schema is ismorphic to the schema for numbers, you could just do:

```hs
declaredNamedSchema _ = 
    s <- declareNamedSchema (Proxy @Int)
    return $ s & name .~ (Just "SecurityParameter")
```

```hs
declareNamedSchema :: proxy a -> Declare (Definitions Schema) NamedSchema
```

`Declare` here is actually a `Monad` with an internal state, with all the already declared schemas.

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +553,209 @@ instance ToSchema (V1 Version) where
+                    & maximum_ .~ Just (fromIntegral (maxBound :: Word64))
+                    & minimum_ .~ Just (fromIntegral (minBound :: Word64))
+                    )
+                )
```
same as above, you could just re-use schemas for `Word16` and `Word64`

```hs
declaredNamedSchema _ = 
    word64Schema <- declareNamedSchema (Proxy @Word64)
    word16Schema <- declareNamedSchema (Proxy @Word16)
    return $ NamedSchema ...
        & properties .~ (mempty
            & at "slot" ?~ word16Schema
            & at "epoch" ?~ word64Schema
```

Or alternatively, rely on the generic instances by using `genericSchemaDroppingPrefix`

```hs
  declareNamedSchema =
    genericSchemaDroppingPrefix "si" (\(--^) props -> props
      & ("slot"  --^ "<description of the slot>")
      & ("epoch" --^ "<description of the epoch>")
```

KtorZ commented on lib/src/Pos/Node/API.hs:
```diff
@@ -547,47 +553,209 @@ instance ToSchema (V1 Version) where
+        object
+            [ "epoch" .= toJSON (Core.getEpochIndex $ Core.siEpoch s)
+            , "slot"  .= toJSON (Core.getSlotIndex  $ Core.siSlot s)
+            ]
```
Can't we rely on `Core.SlotId` already existing instance for this? 

Actually.. that's a very good point. Why are we using `V1` wrapper here :thinking: ? This wrapper exists to avoid orphan instances in the wallet code, but here, we are onto the node so we could just define those instances in the right places, next to the corresponding types!

Anviking replied:
> next to the corresponding types!

Are you saying 1) that we can modify existing JSON instances in Core, 2) that we should do `newtype SlotId = ...` here, or something else?

JSON encoded `Core.SlotId`:
```json
{"siEpoch":4419942607164,"siSlot":{"getSlotIndex":32}}
```

JSON encoded `V1 Core.SlotId`:
```json
"{\"slot\":32,\"epoch\":4419942607164}"
```

Anviking commented on wallet/src/Cardano/Wallet/WalletLayer/Kernel/Settings.hs:
```diff
@@ -20,10 +20,15 @@ import           Paths_cardano_wallet (version)
+        <*> error "TODO"
+        <*> (V1 <$> Node.getFeePolicy node)
+        <*> Node.getSecurityParameter node
   where
```
Should this module really exist? Will we not relay the HTTP requests directly to the node API?

Anviking replied:
Oh… *sees this in timeline* https://github.com/input-output-hk/cardano-wallet/commit/fcffff717f61f57a99a1a1ad2f085cc68db6075a 😄

KtorZ commented:
Looks good :) 

Still some remaining work on the swagger schema though (we ought to get this right by the way, because that's the spec we'll give to the Rust team as well...)

![image](https://user-images.githubusercontent.com/5680256/50681329-66639e00-100b-11e9-9246-d24be87ecdb7.png)

Suggestions: 

- `quantity` for `maxTxSize` can't be negative
- `unknownTag` and `unknownPayload` can be removed. We don't need to "leak" that out of the API. If the `tag` is `"unknown"`, then, there's no `a` and `b` but that could pretty much be it. Actually... we don't even use or want this `TxUnknownPolicy` (which is yet another aberration from the legacy core code). We could simply ignore the `unknown` tag and have the parser failing if anything else is provided. Otherwise, we gonna get some trouble with the round-trip instances. It's also probably a good idea to review the `Arbitrary` instance for `V1 TxFeePolicy` and make sure it doesn't generate any `TxUnknownPolicy`.
- `quantity` for `securityParameter` can't be negative. I still didn't quite get your reserve on the unit here. In the paper, we refer to `k` as `let k = 2 in k slots`, thus, the unit for `securityParameter` is `slot`. `k` refer to a number of slots... ?

Anviking commented:
So I've rebased into 
- auxiliary commits — kind of cherry picks
- purely additive `Extend node API to include protocol parameters`
- Additional refactoring `
Simplify existing instances leveraging helpers`
- [`Fixup: abs in arbitrary, Word instead of Int`](https://github.com/input-output-hk/cardano-sl/pull/3930/commits/9fdb95647c20df4cb5ee79737ee763b12ef9874b) which I'm most unsure about

Compared to last iteration, I've removed max bounds on integers where possible. Partly because `>= 0` looked nicer in the swagger docs, than `[0 .. Int-max-bound-probably]`. Is this fine @KtorZ?

Anviking commented on lib/src/Pos/Node/API.hs:
```diff
@@ -91,6 +95,55 @@ genericSchemaDroppingPrefix prfx extraDoc proxy = do
+                   ToSchema a3,
+                   HasType b (SwaggerType 'Data.Swagger.Internal.SwaggerKindSchema),
+                   IxValue a2 ~ Referenced Schema) =>
+                  UnitOfMeasure -> proxy a3 -> b
```
So re-using this helper in more places isn't super important now. Thought I'd do a separate PR for more drastic simplifications (if any) anyway.*

Still, ultimately would like to do something in this direction:

<details>

```diff
diff --git a/lib/src/Pos/Node/API.hs b/lib/src/Pos/Node/API.hs
index 1a15d4baee..b6d69b7528 100644
--- a/lib/src/Pos/Node/API.hs
+++ b/lib/src/Pos/Node/API.hs
@@ -127,17 +127,11 @@ parseJSONQuantity :: FromJSON a => String -> Value -> Parser a
 parseJSONQuantity s = withObject s $ \o -> o .: "quantity"

 -- assumes there is only one allowed unit
-toSchemaWithUnit :: (HasRequired b [a1], HasProperties b a2,
-                   Monoid b, Monoid a2, At a2, IsString a1, IsString (Index a2),
-                   ToSchema a3,
-                   HasType b (SwaggerType 'Data.Swagger.Internal.SwaggerKindSchema),
-                   IxValue a2 ~ Referenced Schema) =>
-                  UnitOfMeasure -> proxy a3 -> b
 toSchemaWithUnit unitOfMeasure a = (mempty
             & type_ .~ SwaggerObject
             & required .~ ["quantity"]
             & properties .~ (mempty
-                & at "quantity" ?~ toSchemaRef a
+                & at "quantity" ?~ Inline a
                 & at "unit" ?~ (Inline $ mempty
                     & type_ .~ SwaggerString
                     & enum_ ?~ [String $ unitToText unitOfMeasure]
@@ -186,17 +180,9 @@ instance FromJSON LocalTimeDifference where

 instance ToSchema LocalTimeDifference where
     declareNamedSchema _ =
-        pure $ NamedSchema (Just "LocalTimeDifference") $ mempty
-            & type_ .~ SwaggerObject
-            & required .~ ["quantity"]
-            & properties .~ (mempty
-                & at "quantity" ?~ (Inline $ mempty
+        pure $ NamedSchema (Just "LocalTimeDifference")
+             $ toSchemaWithUnit Microseconds $ (mempty
                     & type_ .~ SwaggerNumber
-                    )
-                & at "unit" ?~ (Inline $ mempty
-                    & type_ .~ SwaggerString
-                    & enum_ ?~ ["microseconds"]
-                    )
                 )

 deriveSafeBuildable ''LocalTimeDifference
@@ -724,8 +710,8 @@ instance ToSchema (V1 Core.TxFeePolicy) where
                     & type_ .~ SwaggerString
                     & enum_ ?~ ["linear", "unknown"]
                     )
-                & at "a" ?~ (Inline $ toSchemaWithUnit LovelacePerByte (Proxy @Double))
-                & at "b" ?~ (Inline $ toSchemaWithUnit Lovelace (Proxy @Double))
+                & at "a" ?~ (Inline $ toSchemaWithUnit LovelacePerByte (toSchemaRef $ Proxy @Double))
+                & at "b" ?~ (Inline $ toSchemaWithUnit Lovelace (toSchemaRef $ Proxy @Double))
                 )

 instance Arbitrary (V1 Core.SlotCount) where
```

</details>
